glsl_analyzer: init at 1.4.5

wr7 opened the pull request:
## Description of changes
Created a package for [glsl_analyzer](https://github.com/nolanderc/glsl_analyzer). This is a language server for the OpenGL Shading Language. Currently the only other GLSL language server on Nix is [glslls](https://search.nixos.org/packages?channel=23.11&show=glslls), but it's slightly buggy and has worse descriptions compared to [glsl_analyzer](https://github.com/nolanderc/glsl_analyzer).



I've been using this package for a while, it seems to fully work.

![Package screenshot ](https://github.com/NixOS/nixpkgs/assets/53203261/051d3ae1-663c-4487-bb0c-aa2c81c5c30d)

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wr7 committed:
maintainers: add wr7

wr7 committed:
glsl_analyzer: init at 1.4.5

wr7 commented:
It seems that the Darwin build failed. This seems like an issue with upstream Zig. The only reference that I could find to this is [here](https://github.com/ziglang/zig/issues/10217#issuecomment-1858840393). I'm just going to add Darwin as a badPlatform.

wr7 commented:
As far as testing goes, all of its main features seem to work (hover descriptions, autocomplete, and goto definition).

purcell commented:
> It seems that the Darwin build failed with error: unable to create compilation: DarwinSdkNotFound. This seems like an issue with upstream Zig. The only reference that I could find to this is https://github.com/ziglang/zig/issues/10217#issuecomment-1858840393. I'm just going to add Darwin as a badPlatform.

Zig in nixpkgs fails to build on aarch64-darwin with exactly that error when the sandbox is enabled. I found this locally in an unrelated setting and disabling the sandbox fixed it.

purcell commented:
> It seems that the Darwin build failed with error: unable to create compilation: DarwinSdkNotFound. This seems like an issue with upstream Zig. The only reference that I could find to this is https://github.com/ziglang/zig/issues/10217#issuecomment-1858840393. I'm just going to add Darwin as a badPlatform.

Zig in nixpkgs fails to build on aarch64-darwin with exactly that error when the sandbox is enabled. I found this locally in an unrelated setting and disabling the sandbox fixed it.

wr7 commented:
> > It seems that the Darwin build failed with error: unable to create compilation: DarwinSdkNotFound. This seems like an issue with upstream Zig. The only reference that I could find to this is [ziglang/zig#10217 (comment)](https://github.com/ziglang/zig/issues/10217#issuecomment-1858840393). I'm just going to add Darwin as a badPlatform.
> 
> Zig in nixpkgs fails to build on aarch64-darwin with exactly that error when the sandbox is enabled. I found this locally in an unrelated setting and disabling the sandbox fixed it.

So is there any way to fix this for the package? I got that error from OFBorg. I would assume that the Hydra would have the same error when building the package for Darwin.

purcell commented:
I didn't figure out what the fix might be, sorry, and didn't keep the error around long enough to file a primary bug against the `zig` derivation.

wr7 commented:
It seems the same issue was happening to `zls`. Going off of the pull request there (#290196), I've likely fixed the issue.

wr7 commented:
Zig 12 was recently released, but it makes sense to still use Zig 11 because using Zig 12 requires more intensive patching.

wr7 commented:
Upstream now compiles with Zig 12 (https://github.com/nolanderc/glsl_analyzer/pull/56). I've just updated the package to use that instead of my Zig 11 patch.

I also rebased.

Aleksanaa commented on pkgs/by-name/gl/glsl_analyzer/package.nix:
```diff
@@ -0,0 +1,36 @@
+  pname = "glsl_analyzer";
+  version = "1.4.4";
+
+  src = fetchgit {
```
Please use `fetchFromGitHub` instead. Also is `leaveDotGit` necessary?

wr7 replied:
I've switched to `fetchFromGitHub`, but glsl_analyzer does not compile without `leaveDotGit` because it reads the commit hash for use in `glsl_analyzer --version`. This is also why `gitMinimal` is in `nativeBuildInputs`.

Aleksanaa commented on pkgs/by-name/gl/glsl_analyzer/package.nix:
```diff
@@ -0,0 +1,36 @@
+  ];
+
+  meta = {
+    description = "Language server for GLSL (OpenGL Shading Language).";
```
```suggestion
    description = "Language server for GLSL (OpenGL Shading Language)";
```

Aleksanaa commented:
Could we add something like

```
postPatch = ''
  substituteInPlace build.zig \
    --replace-fail 'b.run(&.{ "git", "describe", "--tags", "--always" })' '${src.rev}'
'';
```

To replace https://github.com/nolanderc/glsl_analyzer/blob/da890897ef5c4b1e6328cd614ed1808f00a0d351/build.zig#L100

Aleksanaa commented:
leaveDotGit is potentially not deterministic so we'd better avoid it

wr7 commented:
> leaveDotGit is potentially not deterministic so we'd better avoid it

Good to know. I've just applied your suggestion, and `glsl_analyzer --version` still works.

Aleksanaa commented on pkgs/by-name/gl/glsl_analyzer/package.nix:
```diff
@@ -0,0 +1,41 @@
+
+  nativeBuildInputs = [
+    zig_0_12.hook
+    gitMinimal
```
```suggestion
```

Not needed anymore?

wr7 commented:
I also realized that the commit description was no-longer accurate. I just removed it.

wr7 commented:
A new version was just released. It makes sense to just use that instead.

pluiedev commented:
Result of `nixpkgs-review pr 295076` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>glsl_analyzer</li>
  </ul>
</details>
